### PR TITLE
Implemented Bid API

### PR DESCRIPTION
**Pull request by fvukelic**
Wrote some additional tests for the Bid API and added the necessary endpoints.
- TG-66
- TG-67
- TG-68
- TG-69

### CODE REVIEW

**Comment by tzumainn** [comment]
@fvukelic can you rebase on top of master to pull in the changes from https://github.com/CCI-MOC/flocx-market/pull/33 ? That should pull in the updates to have the rest api run from a script defined by an entrypoint. You can copy that pattern to create a script for initializing the database, similar to https://github.com/openstack/ironic/blob/master/ironic/cmd/dbsync.py

File: flocx_market/db/sqlalchemy/bid_api.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        self.marketplace_bid_id = str(uuid4())
+        self.creator_bid_id = creator_bid_id
+        self.creator_id = creator_id
+        self.server_quantity = server_quantity
+        self.start_time = start_time
+        self.end_time = end_time
+        self.duration = duration
+        self.status = status
+        self.server_config_query = server_config_query
+        self.cost = cost
```
Comment (larsks) [inline flocx_market/db/sqlalchemy/bid_api.py]: Let's drop this `__init__` method here just like in #e49134845c4d754f01b1c5d5fdd860cf27fccca4. The one provided by the underlying `*Model` object is more useful.

File: flocx_market/tests/base_test.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
+from unittest import TestCase
```
Comment (larsks) [inline flocx_market/tests/base_test.py]: Do we actually need this file?

File: flocx_market/api/app.py [hunk 3]
```diff
@@ -1,7 +1,10 @@
 from flask import Flask
 from flask_restful import Api
 from flocx_market.api.offer import Offer
+from flocx_market.api.bid import Bid
 from flocx_market.api.root import Root
+from flask import Blueprint
+from flask import jsonify
```
Comment (tzumainn) [inline flocx_market/api/app.py]: I don't think we need these two imports

**Comment by larsks** [comment]
If you rebase instead of merge, you'll get rid of that extra commit and the commit checks will pass (leaving you with the pep8 failures).

File: flocx_market/api/app.py (hunk 3 above)
Comment (larsks) [reply flocx_market/api/app.py]: @tzumainn Those now show up in the CI tests (since we merged 7ccb4892754532014286da63c5e83b5a6bd0e064).

Comment (fvukelic) [reply flocx_market/api/app.py]: And now the tests are failing :D

Comment (fvukelic) [reply flocx_market/api/app.py]: Someone create a test_offer_api that is failing @tzumainn @larsks

Comment (tzumainn) [reply flocx_market/api/app.py]: Odd, it's not failing for me (or for the CI) - is it possible there's some sort of merge conflict?

Comment (fvukelic) [reply flocx_market/api/app.py]: @tzumainn I fixed it, now it passes

Comment (fvukelic) [reply flocx_market/api/app.py]: also I think if it is communicating with the db, it should be an integration test and not a unit test.

File: flocx_market/tests/unit/models/test_offer_unit.py [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+                      "ram_gb": 512, "storage_type": "samsung SSD",
+                      "storage_size_gb": 204}, 11.5)
+        self.assertEqual(o.provider_id, "b9752cc0-9bed-4f1c-8917-12ade7a6fdbe")
+        self.assertEqual(o.creator_id, "Alice1992")
+        self.assertEqual(o.marketplace_date_created, now)
+        self.assertEqual(o.status, "available")
+        self.assertEqual(o.server_id, "Dell380")
+        self.assertEqual(o.start_time, now)
+        self.assertEqual(o.end_time, now)
+        self.assertEqual(o.cost, 11.5)
```
Comment (tzumainn) [inline flocx_market/tests/unit/models/test_offer_unit.py]: I think Lars moved this file to tests/unit/db/sqlalchemy/test_offer_api.py, so there's no need to reintroduce this here.

File: flocx_market/api/app.py (hunk 3 above)
Comment (tzumainn) [reply flocx_market/api/app.py]: Agreed. None of the app tests are communicating with the db, are they?

Comment (fvukelic) [reply flocx_market/api/app.py]: @tzumainn in tests/unit/db/sqlalchemy/test_offer_api.py all of them are

File: flocx_market/tests/unit/models/test_offer_unit.py (hunk 4 above)
Comment (fvukelic) [reply flocx_market/tests/unit/models/test_offer_unit.py]: That file is different, calling db method. 
This one is only testing the model creation without communicating to the db

File: flocx_market/tests/unit/models/test_offer_unit.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
+from flocx_market.tests.unit.unit_base_test import UnitBaseTest
```
Comment (larsks) [inline flocx_market/tests/unit/models/test_offer_unit.py]: Please take this out of the pull request.  This pull request is for implementing the Bid api, and should not have any offer related funcions.

File: flocx_market/tests/unit/models/test_offer_unit.py (hunk 4 above)
Comment (larsks) [reply flocx_market/tests/unit/models/test_offer_unit.py]: This is  identical to https://github.com/CCI-MOC/flocx-market/blob/master/flocx_market/tests/unit/db/sqlalchemy/test_offer_api.py#L23

File: flocx_market/api/app.py (hunk 3 above)
Comment (tzumainn) [reply flocx_market/api/app.py]: Ah, but that's not an app test (testing the REST API). That's testing the database API.

File: flocx_market/tests/base_test.py (hunk 2 above)
Comment (fvukelic) [reply flocx_market/tests/base_test.py]: deleted it

File: flocx_market/tests/unit/models/test_offer_unit.py (hunk 4 above)
Comment (fvukelic) [reply flocx_market/tests/unit/models/test_offer_unit.py]: sounds good, it was a confusion on my side then on what needs to go into the unit tests @tzumainn @larsks

Comment (fvukelic) [reply flocx_market/tests/unit/models/test_offer_unit.py]: I will add a bunch of more tests that are similar to the test_offer_api for bids and contracts

File: flocx_market/db/sqlalchemy/bid_api.py (hunk 1 above)
Comment (fvukelic) [reply flocx_market/db/sqlalchemy/bid_api.py]: the only thing with removing this is that then the marketplace_offer_id is expected as an argument. If we keep it in there, we generate the marketplace_offer_id as a UUID during the init method. How would you suggest resolving that if we remove it? @larsks

Comment (tzumainn) [reply flocx_market/db/sqlalchemy/bid_api.py]: Something like this: https://github.com/CCI-MOC/esi-leap/blob/master/esi_leap/db/sqlalchemy/api.py#L166

Comment (fvukelic) [reply flocx_market/db/sqlalchemy/bid_api.py]: I'm not a 100% sure if removing one method to add another one is the move? and I don't know how to do the UUID generation without it

Comment (tzumainn) [reply flocx_market/db/sqlalchemy/bid_api.py]: Having a specific `create` method that also lets you set defaults makes a lot of sense to me. It seems odd to have the `__init__` method of BidApi be the equivalent of creating a new Bid.

UUID generation can be handled through oslo utils, as seen in the linked code. In general, whenever there's an oslo library that we can use, we should *always* be using it.

Comment (fvukelic) [reply flocx_market/db/sqlalchemy/bid_api.py]: oh okay that makes sense @tzumainn. I just realized that Lars used a different approach here: https://github.com/CCI-MOC/flocx-market/commit/e49134845c4d754f01b1c5d5fdd860cf27fccca4

what do you think?

Comment (fvukelic) [reply flocx_market/db/sqlalchemy/bid_api.py]: @larsks sorry for the confusion there, I completely missed that you were generating the UUID on the model itself. It makes way more sense that way

Comment (tzumainn) [reply flocx_market/db/sqlalchemy/bid_api.py]: @fvukelic I'd personally prefer for it to get in the api, and not model. We're also going to end up doing things such as defaulting project_id (taken from the context) and the like. But I don't have a strong opinion on whether that has to be done right now.

File: flocx_market/api/app.py [hunk 6]
```diff
@@ -16,6 +17,9 @@ def create_app(app_name):
     )
     app.config['PROPAGATE_EXCEPTIONS'] = CONF.flask.PROPAGATE_EXCEPTIONS
     api = Api(app)
-    api.add_resource(Offer, '/offer', '/offers/<string:marketplace_offer_id>')
+    api.add_resource(Offer, '/offer', '/offers/',
+                     '/offers/<string:marketplace_offer_id>')
+    api.add_resource(Bid, '/bid', '/bids/',
+                     '/bids/<string:marketplace_bid_id>')
```
Comment (larsks) [inline flocx_market/api/app.py]: The specification requires that we use `/bid` and `/bid/<id>`, not `/bids/<id>`.

File: flocx_market/api/bid.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
+from flask_restful import Resource
+from flocx_market.db.sqlalchemy.bid_api import BidApi
+from flask import request
+
+
+class Bid(Resource):
+
+    @classmethod
+    def get(cls, marketplace_bid_id):
```
Comment (larsks) [inline flocx_market/api/bid.py]: This will fail for `GET /bid` because `marketplace_bid_id` is a required parameter.  Change it to a keyword parameter with a default value:

```
def get(cls, marketplace_bid_id=None):
```

Also, this is why we write tests :). The error shows up very quickly if you test the behavior against the specification.

File: flocx_market/api/bid.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    @classmethod
+    def delete(cls, marketplace_bid_id):
+        bid = BidApi.find_by_id(marketplace_bid_id)
+        if bid:
+            bid.delete_from_db()
+            return {'message': 'Bid deleted.'}
+        return {'message': 'Bid not found.'}, 404
+
+    @classmethod
+    def put(cls, marketplace_bid_id):
```
Comment (larsks) [inline flocx_market/api/bid.py]: This should return a 404 error if the client attempts to `PUT` against an offer that doesn't exist.

File: flocx_market/api/offer.py [hunk 9]
```diff
@@ -6,7 +6,10 @@
 class Offer(Resource):
 
     @classmethod
-    def get(cls, marketplace_offer_id):
+    def get(cls, marketplace_offer_id=None):
```
Comment (larsks) [inline flocx_market/api/offer.py]: As discussed in #64, let's put this in a separate PR.

File: flocx_market/tests/unit/common/test_service.py [hunk 10]
```diff
@@ -39,12 +39,12 @@ def test_service_cli_valid(find_config_files):
            str(args[args.index('--api-public_endpoint') + 1]))
     assert(CONF.api.api_workers ==
            int(args[args.index('--api-api_workers') + 1]))
-    assert(CONF.api.enable_ssl_api)
+    assert CONF.api.enable_ssl_api
```
Comment (larsks) [inline flocx_market/tests/unit/common/test_service.py]: This doesn't appear to be related to the Bid API.

File: flocx_market/tests/unit/db/sqlalchemy/test_offer_api.py [hunk 11]
```diff
@@ -1,20 +1,19 @@
 import datetime
```
Comment (larsks) [inline flocx_market/tests/unit/db/sqlalchemy/test_offer_api.py]: Ditto for this; let's move non-Bid stuff to separate PRs.

File: setup.cfg [hunk 12]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (larsks) [inline setup.cfg]: I don't think we need this change.

File: tox.ini [hunk 13]
```diff
@@ -26,10 +26,10 @@ commands =
 
 [testenv:py37]
 commands = 
-        pytest --cov=flocx_market {posargs:flocx_market/tests/unit}
+        pytest {posargs:flocx_market/tests/unit}
```
Comment (larsks) [inline tox.ini]: Please remove these changes from the PR as well.  Thanks!

File: flocx_market/db/sqlalchemy/bid_api.py (hunk 1 above)
Comment (larsks) [reply flocx_market/db/sqlalchemy/bid_api.py]: @tzumainn I kind of want default values in the model. Let me know if you have any free time tomorrow morning and we can chat?

File: flocx_market/api/bid.py (hunk 8 above)
Comment (larsks) [reply flocx_market/api/bid.py]: (I've fixed this for offers in #65)

File: flocx_market/db/sqlalchemy/bid_api.py (hunk 1 above)
Comment (tzumainn) [reply flocx_market/db/sqlalchemy/bid_api.py]: @larsks Sure! I've seen some interesting things in the OpenStack code, and would not mind your opinion on them.

File: flocx_market/api/app.py (hunk 6 above)
Comment (fvukelic) [reply flocx_market/api/app.py]: I know, I talked to @SahilTikale about changing that to follow the REST conventions. Which do you think is the right approach @larsks?

Comment (larsks) [reply flocx_market/api/app.py]: I would implement it according to the spec right now, and we can talk with Sahil when he's back if you think we should update the spec.

Comment (fvukelic) [reply flocx_market/api/app.py]: Sounds good

**Comment by fvukelic** [comment]
can we merge this @larsks @tzumainn?

**Comment by fvukelic** [comment]
looks like there was a new merge and there's a conflicting file now, I'll change that fast and resubmit @larsks @tzumainn, thank for the approve!

**Comment by larsks** [comment]
@fvukelic just noticed this still has a `/bids` endpoint, which we had talked about removing since it's contrary to the spec document.

**Comment by fvukelic** [comment]
> @fvukelic just noticed this still has a `/bids` endpoint, which we had talked about removing since it's contrary to the spec document.

hmm that is really weird, I don't have it locally, I'll remove it @larsks

**Comment by larsks** [comment]
Maybe I was looking at an older version of the pr (I was looking locally)?  If so, sorry for the noise!

**Comment by fvukelic** [comment]
The only problem I'm seeing is that I have each model and model api separated into separate files, but @tzumainn merged a PR with having them all in a single file, so this is causing a merge conflict. Which approach should we take here @tzumainn @larsks? When I asked 2-3 weeks ago, you said it was fine if we keep them separated, let me know if this changed @tzumainn and I'll modify my PR. Thanks!

**Comment by larsks** [comment]
I don't think we need to update this PR with those changes yet. Just resolve the conflict in `flocx_market/db/sqlalchemy/offer_api.py` (by deleting if after a rebase), and I think we're good to merge.

**Comment by fvukelic** [comment]
@larsks, just rebased and Travis likes it
